wagtail/admin/edit_handlers.py

```diff
@@ -789,8 +792,89 @@ def render(self):
         )
 
 
+class CommentPanel(EditHandler):
+    def __init__(self, *args, **kwargs):
+        self.comments = []
+        super().__init__(*args, **kwargs)
+
+    def required_formsets(self):
+        # add the comments formset
+        # we need to pass in the current user for validation on the formset
+        # this could alternatively be done on the page form itself if we added the
+        # comments formset there, but we typically only add fields via edit handlers
+        current_user = getattr(self.request, 'user', None)
+
+        class CommentReplyFormWithRequest(CommentReplyForm):
+            user = current_user
+
+        class CommentFormWithRequest(CommentForm):
+            user = current_user
+
+            class Meta:
+                formsets = {
+                    'replies': {
+                        'form': CommentReplyFormWithRequest
+                    }
+                }
+
+        return {
+            'comments': {
+                'form': CommentFormWithRequest,
+                'fields': ['text', 'contentpath'],
+            }
+        }
+
+    template = "wagtailadmin/edit_handlers/comments/comment_panel.html"
+    js_template = "wagtailadmin/edit_handlers/comments/comment_panel.js"
+    declarations_template = "wagtailadmin/edit_handlers/comments/comment_declarations.html"
+
+    def html_declarations(self):
+        return render_to_string(self.declarations_template)
+
+    def on_form_bound(self):
+        self.comments = []
+        comment_forms = self.form.formsets['comments'].forms
+        for form in comment_forms:
+            instance = form.instance
+            instance.replies = [reply_form.instance for reply_form in form.formsets['replies'].forms]
+            self.comments.append(instance)
+
+    def get_context(self):
+        user = getattr(self.request, 'user', None)
+        authors = {user.pk: user_display_name(user)} if user else {}
+        serialized_comments = []
+        for comment in self.comments:
+            # iterate over comments to retrieve user display names and serialized versions
+            if comment.user_id not in authors:
+                authors.update({comment.user.pk: user_display_name(comment.user)})
+            serialized_comments.append(comment.serializable_data())
+            for reply in comment.replies.all():
```

Comment:
**Thought:** Could do a simple optimisation? Not sure how well this works with modelcluster in the context you are here in.

```suggestion
            for reply in comment.replies.select_related('user'):
```

_This would be redundant if you prefetch it level higher, then you must use `all`._

I am not sure how possible it is to prefetch replies if we are not fetching all of this stuff from a database table, but it might be a nice optimisation.